### PR TITLE
RDCC-1660: PBA Changes

### PR DESCRIPTION
**Pull request by AdilOoze**
### JIRA link (if applicable) ###

Development: https://tools.hmcts.net/jira/browse/RDCC-1660
QA/Deployment: https://tools.hmcts.net/jira/browse/RDCC-1849

### Change description ###

PBA related changes are atomic and transaction Boundary defined

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/PaymentAccountServiceImpl.java [hunk 1]
```diff
@@ -83,12 +82,13 @@ public void deleteUserAccountMaps(Organisation organisation) {
     }
 
 
+    @Transactional
     public void deletePaymentAccountsFromOrganisation(Organisation organisation) {
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/PaymentAccountServiceImpl.java]: @transactional annotation should not be added here. It should be above editPaymentAccountsByOrganisation

File: src/main/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalController.java [hunk 2]
```diff
@@ -225,20 +225,19 @@ public ResponseEntity editPaymentAccountsByOrgId(@Valid @NotNull @RequestBody Pb
...
                                                      @PathVariable("orgId") @NotBlank String organisationIdentifier) {
+
         log.info("{}:: Received request to edit payment accounts by organisation Id...", loggingComponentName);
 
         paymentAccountValidator.validatePaymentAccounts(pbaEditRequest.getPaymentAccounts(), organisationIdentifier);
         Optional<Organisation> organisation = Optional.ofNullable(organisationService
                 .getOrganisationByOrgIdentifier(organisationIdentifier));
 
-        if (!organisation.isPresent()) {
+        if (organisation.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (AdilOoze) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalController.java:234]: Addressed here: https://github.com/hmcts/rd-professional-api/pull/731/commits/85c67034e07347859366c191a92af4c162d7a4ee

File: src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/PaymentAccountServiceImpl.java [hunk 3]
```diff
@@ -37,75 +40,51 @@
...
 
         /** Please Note:
          * The below lines are required to set the Organisation's Payment Accounts List to be empty.
          * If this is not done, the Organisation's list will still contain the previous Accounts */
-        List<PaymentAccount> resetOrganisationPaymentAccounts = new ArrayList<>();
-        organisation.setPaymentAccounts(resetOrganisationPaymentAccounts);
+        organisation.setPaymentAccounts(new ArrayList<>());
     }
 
     public void addPaymentAccountsToOrganisation(PbaEditRequest pbaEditRequest, Organisation organisation) {
```
[the remark this answers is not in the document]
Comment (AdilOoze) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/PaymentAccountServiceImpl.java:90]: Addressed here: https://github.com/hmcts/rd-professional-api/pull/731/commits/a78336964218a809136399438cd89e4277d5aa21

[the remark this answers is not in the document]
Comment (AdilOoze) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/PaymentAccountServiceImpl.java:91]: Addressed here: https://github.com/hmcts/rd-professional-api/pull/731/commits/0b759ac75882cb143a9e079f8821b0d268f10dda

File: src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/PaymentAccountServiceImpl.java (hunk 1 above)
Comment (AdilOoze) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/PaymentAccountServiceImpl.java]: Addressed here: https://github.com/hmcts/rd-professional-api/pull/731/commits/ab5e48e12c529ac0a7a065a4eb4a4bfaa1ca8c1f

File: src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/PaymentAccountServiceImpl.java [hunk 4]
```diff
@@ -120,14 +99,25 @@ public PbaResponse addUserAndPaymentAccountsToUserAccountMap(Organisation organi
...
                 .map(account -> new UserAccountMapId(user, account))
                 .collect(Collectors.toList());
     }
-}
+
+    public void removePaymentAccountAndUserAccountMaps(List<UserAccountMapId> userAccountMaps) {
+        EntityManager em = emf.createEntityManager();
+        EntityTransaction transaction = em.getTransaction();
+        transaction.begin();
+
```
[the remark this answers is not in the document]
Comment (AdilOoze) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/PaymentAccountServiceImpl.java]: Addressed here: https://github.com/hmcts/rd-professional-api/pull/731/commits/16a18ef00e8d4edeeebe0b9ad365880afeecd9d3

**Review by shashank-rastogi — APPROVED** [review]
Adil. - the changes looks fine.. I have approved.. but wanted to ask if this requires an explicit transaction begin and commit. Can you remove and try if it works?

**Review by shashank-rastogi — APPROVED** [review]

**Review by shashank-rastogi — APPROVED** [review]

**Review by pksaxena01 — APPROVED** [review]
Approved

**Pull request merged**